### PR TITLE
[JUJU-1214] Remove Windows charm support from Juju for 3.0

### PR DESCRIPTION
**Pull request by barrettj12**
Remove Windows support from Juju as an agent.

With the release of Juju 3.0, Juju will know longer support deploying
windows based charms. This commit does the work to remove all Windows
references and code from the code base related to deploying Windows
charms.

Juju will still support Windows for the client and as such some Windows
references have left behind as the client has references to agent based
code.

Where appropriate all tests have been kept or altered to another series
such that test coverage is not affected.

## QA steps

The Windows client should still compile:
```
GOOS=windows go build -a github.com/juju/juju/cmd/juju
```
All the CI below should pass as well (except the ones with known failures).

We can't add Windows machines:
```console
$ juju add-machine --series win7
$ juju status
...
Machine  State  Inst id  Series  AZ  Message
...
1        down   pending  win7        unknown OS for series: "win7"
...
```

## To keep

To support the Windows client, (I believe) we need to keep Windows stuff in:
- `caas/kubernetes/provider/exec/termsize_windows.go`
- `cmd/plugins`
- `core/os`
- `core/paths`
- `juju/osenv`
- `juju/sockets`
- `service/systemd`
- `state/backups`
- `wrench`

We should also keep historical version numbers containing Windows (e.g. `1.2.3-windows-alpha`)

## Changes

- Deleted:
  - `acceptancetests/repository/charms-win/`
  - (list to be completed)
- Removed all Windows stuff in:
  - `apiserver/`
  - `cloudconfig/`
  - Several tests in `cmd/juju/commands/`
  - `cmd/juju/machine/add.go`
  - `cmd/juju/romulus/agree/agree_test.go`
  - Some tests in `cmd/juju/ssh/`
  - `cmd/jujuc/`
  - `cmd/jujud/`
  - `cmd/containeragent/`
  - `provider/`
  - `state/`
  - `worker/`
  - (list to be completed)
- Changed the install targets in Makefile so that on Windows, it will only try to install the client.
- Removed agent binary names from `juju/names/names_windows`

## Questions

I'm not sure if I should:
- Remove Windows from:
  - `agent/` - is this used by the client?
  - `caas/kubernetes/provider/exec/termsize_windows.go`
  - `environs/tools/simplestreams_test.go`

### CODE REVIEW

**Comment by juanmanuel-tirado** [comment]
I love this branch :)

I added the "do-not-merge" label. Feel free to remove it.

**Comment by barrettj12** [comment]
So something's not right. I removed the agent names from `juju/names/names_windows.go`, and now when you try
```
GOOS=windows go build github.com/juju/juju/cmd/juju
```
it complains about
```
# github.com/juju/juju/agent/tools
agent/tools/symlinks.go:43:44: undefined: names.Jujuc
agent/tools/symlinks.go:46:45: undefined: names.Jujud
```
This looks like a case of bad dependency - the client is importing `X` from package `Y`, and somewhere else package `Y` imports (something which imports) `agent/tools`. I can't find the path from `cmd/juju` to `agent/tools` though.

**Comment by barrettj12** [comment]
Found it:
```
github.com/juju/juju/cmd/juju
github.com/juju/juju/cmd/juju/commands
github.com/juju/juju/environs
github.com/juju/juju/cloudconfig/instancecfg
github.com/juju/juju/agent/tools
```
... this looks like a mess.

File: state/application_test.go
Comment (wallyworld) [inline state/application_test.go:680]: Use centos or something but need to keep the test

File: worker/uniter/runner/args.go
Comment (wallyworld) [inline worker/uniter/runner/args.go:37]: can look to remove this fucntion now, it doesn't add anything

File: worker/uniter/runner/jujuc/util_test.go
Comment (wallyworld) [inline worker/uniter/runner/jujuc/util_test.go:35]: delete this function

**Review by wallyworld — CHANGES_REQUESTED** [review]
Just a few small things, the main issue seems to be where we had logic for !windows that has been removed by mistake.
I may have missed it, but we also want logic (if not there already) to prevent Windows charms from being deployed - short circuit at deploy time. Also need to prevent `add-machine --series=win2012` and fail early (if not already done).

File: cmd/juju/ssh/debughooks_test.go
Comment (barrettj12) [reply cmd/juju/ssh/debughooks_test.go:111]: It's already closed

File: juju/names/names.go
Comment (barrettj12) [reply juju/names/names.go:5]: Nah the binary names are needed on Windows for the client to compile (because of bad dependencies). This is the issue I was talking about earlier

File: cmd/jujud/reboot/reboot.go
Comment (tlm) [inline cmd/jujud/reboot/reboot.go:170]: We're setting the shebang line here to bash but we're executing it with bourne shell below. Should be a little more explicit.

File: cmd/jujud/reboot/reboot.go
Comment (tlm) [inline cmd/jujud/reboot/reboot.go:186]: We're executing the file the actual interpreter so we don't need this script to be executable in fact it's probably better from a security stand point.

File: cmd/jujud/run/run_test.go
Comment (tlm) [inline cmd/jujud/run/run_test.go:376]: This should be path.Join()

File: common/sender/sender.go
Comment (tlm) [inline common/sender/sender.go:1]: Update copyright year

File: environs/imagemetadata/urls_test.go
Comment (tlm) [inline environs/imagemetadata/urls_test.go:46]: Don't think this should have `/images` on the end.

File: environs/tools/urls_test.go
Comment (tlm) [inline environs/tools/urls_test.go:149]: Think this shouldn't have /tools on it.

**Comment by tlm** [comment]
@wallyworld want to take another look over this before we look at hitting the merge button

**Review by wallyworld — APPROVED** [review]

**Pull request merged**
